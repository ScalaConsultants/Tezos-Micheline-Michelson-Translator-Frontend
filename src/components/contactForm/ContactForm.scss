@import "../../scss/vars";

.contact-form {
    .contact-form_line {
        display: flex;
        justify-content: space-between;

        @media screen and (max-width: 500px) {
            display: block;
        }
    }

    .contact-form_button {
        font-size: 16px;
        margin-top: 15px;
        background: $background-color-red;
        color: white;
        box-shadow: $box-shadow;
        border-radius: 4px;
        height: 40px;
        width: 100px;
        text-transform: uppercase;
        display: inline-block;

        &:hover {
            cursor: pointer;
        }
    }

    .form-input {
        width: 80%;

        .form-input_input {
            width: 90%;
        }

        @media screen and (max-width: 500px) {
            width: 100%;

            .form-input_input {
                width: 95%;
            }
        }
    }

    .contact-form_message {
        width: 100%;

        .form-input_input {
            width: 95%;
        }

        .form-input_validation-marker {
            left: calc(92% + 3px);

            @media screen and (max-width: 500px) {
                left: 85%;
            }
        }
    }

    .contact-form_name {
        width: 100%;

        .form-input_input {
            width: 45%;
            @media screen and (max-width: 500px) {
                width: 95%;
            }
        }

        .form-input_validation-marker {
            left: calc(42% + 3px);

            @media screen and (max-width: 500px) {
                left: 85%;
            }
        }
    }
}
