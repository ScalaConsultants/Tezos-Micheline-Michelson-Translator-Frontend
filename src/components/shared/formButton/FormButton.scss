@import "../../../scss/vars";

.form-button {
    font-weight: 500;
    text-transform: uppercase;
    letter-spacing: 0.15;
    font-size: 14px;
    box-shadow: $box-shadow;
    border-radius: 4px;
    padding: 5px;
    height: 40px;
    min-width: 100px;
    border: none;
    color: white;

    &:hover {
        cursor: pointer;
    }
}

.form-button_submit {
    background: $background-color-red;
}

.form-button_secondary {
    background: $background-color-grey-3;
}
