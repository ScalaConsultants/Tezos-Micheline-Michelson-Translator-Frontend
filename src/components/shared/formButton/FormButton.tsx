import React from "react";
import { ButtonProps } from "./types";

const FormButton = ({ type, label, className, disabled, stylingType, onClick }: ButtonProps) => (
  <button
    className={`form-button form-button_${stylingType || "submit"} ${className}`}
    data-testid="form-button"
    disabled={disabled}
    type={type}
    onClick={onClick}
  >
    {label}
  </button>
);

export default FormButton;
