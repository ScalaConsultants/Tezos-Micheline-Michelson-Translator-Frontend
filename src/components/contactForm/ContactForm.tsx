import React, { useEffect, useState } from "react";
import { useDispatch, useMappedState } from "redux-react-hook";
import { Formik } from "formik";
import { useGoogleReCaptcha } from "react-google-recaptcha-v3";
import FormInput from "../shared/input/FormInput";
import { FormValues } from "./types";
import "./ContactForm.scss";
import * as MessageTypes from "../../store/message/types";
import * as MessageActions from "../../store/message/actions";
import Alert from "../shared/alert/Alert";
import FormButton from "../shared/formButton/FormButton";
import { removeEmptyProperties } from "../../helpers/tools";
import { bindActionCreators } from "redux";

const mapState = (state: MessageTypes.IMessageGlobalState) => ({
  message: state.message
});

const ContactForm = () => {
  const dispatch = useDispatch();
  const { message } = useMappedState(mapState);
  const { executeRecaptcha } = useGoogleReCaptcha();
  const boundMessageActions = bindActionCreators(MessageActions, dispatch);
  const minInputLengths = {
    minPhoneNumberLength: 9,
    minEmailAddressLength: 6,
    minNameLength: 3
  };
  const [status, setStatus] = useState({success: false});

  const submitForm = async (values: FormValues) => {
    if (!executeRecaptcha) return;

    const token = await executeRecaptcha("contact_form");
    if (!token.length) return;

    const data = removeEmptyProperties(values);

    boundMessageActions.MessageSet(data);
    boundMessageActions.MessageSend(data, token);
    setStatus({success: true});
  };

  useEffect(() => {
    boundMessageActions.MessageSetError(null);
  }, []);

  const saveForm = (values: FormValues) => {
    const data = removeEmptyProperties(values);
    boundMessageActions.MessageSet(data);
  };

  type validationErrors = {
    name?: string;
    phone?: string;
    email?: string;
    content?: string;
  };

  const validate = (values: FormValues) => {
    saveForm(values);

    const errors: validationErrors = {};

    if (!values.name) {
      errors.name = "Required";
    } else if (values.name.length < minInputLengths.minNameLength) {
      errors.name = "Name is too short";
    }

    if (!values.phone && !values.email) {
      errors.phone = "You need to provide an email or a phone number";
    } else if (
      values.phone &&
      !/^[+]*[(]{0,1}[0-9]{1,4}[)]{0,1}[-\s./0-9]*$/i.test(values.phone)
    ) {
      errors.phone = "It's a wrong phone number!";
    } else if (values.phone.length < minInputLengths.minPhoneNumberLength) {
      errors.phone = "Phone number is to short";
    }

    if (!values.email && !values.phone) {
      errors.email = "You need to provide an email or a phone number";
    } else if (
      (values.email &&
        !/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/i.test(values.email)) ||
      values.email.length < minInputLengths.minEmailAddressLength
    ) {
      errors.email = "It's a wrong email address!";
    }

    if (!values.content) {
      errors.content = "Required";
    }

    return errors;
  };

  return (
    <div className="contact-form">
      <h2>Send message</h2>
      <Formik
        initialValues={{
          name: message.name,
          phone: message.phone,
          email: message.email,
          content: message.content
        }}
        onSubmit={submitForm}
        validate={validate}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleBlur,
          handleSubmit,
          isSubmitting
        }) => (
          <form onSubmit={handleSubmit} noValidate>
            <div className="contact-form_line">
              <FormInput
                label="Name"
                type="text"
                name="name"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.name}
                errors={errors.name}
                touched={touched.name}
                className="contact-form_name"
              />
            </div>
            <div className="contact-form_line">
              <FormInput
                label="Phone"
                type="text"
                name="phone"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.phone}
                errors={errors.phone}
                touched={touched.phone}
              />
              <FormInput
                label="Email"
                type="email"
                name="email"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.email}
                errors={errors.email}
                touched={touched.email}
              />
            </div>
            <div className="contact-form_line">
              <FormInput
                label="How we can help you?"
                type="text"
                name="content"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.content}
                errors={errors.content}
                touched={touched.content}
                className="contact-form_message"
              />
            </div>
            <FormButton
              label="Submit"
              stylingType="submit"
              type="submit"
              disabled={isSubmitting}
            />
            {status.success && <Alert type="success" message="Message sent." />}
          </form>
        )}
      </Formik>
    </div>
  );
};

export default ContactForm;
