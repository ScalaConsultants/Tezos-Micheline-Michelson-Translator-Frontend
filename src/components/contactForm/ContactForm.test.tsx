import React from "react";
import { mount, shallow } from "enzyme";
import { StoreContext } from "redux-react-hook";
import ContactForm from "./ContactForm";
import FormButton from "../shared/formButton/FormButton";

import configureStore from "../../store";
import rootSaga from "../../store/saga";

const store = configureStore();
store.runSaga(rootSaga);

describe("ContactForm", () => {
  const wrapper = mount(
    <StoreContext.Provider value={store}>
      <ContactForm />
    </StoreContext.Provider>,
  );

  it("matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders without crashing", () => {
    const item = wrapper.find(ContactForm);

    expect(
      item
        .at(0)
        .find({ name: "name" })
        .at(0)
        .text(),
    ).toEqual("Name");

    expect(
      item
        .at(0)
        .find(FormButton)
        .at(0)
        .text(),
    ).toEqual("Submit");
  });
});
