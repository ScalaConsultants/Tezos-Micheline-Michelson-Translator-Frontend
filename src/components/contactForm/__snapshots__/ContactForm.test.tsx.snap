// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm matches snapshot 1`] = `
"<ContactForm>
  <div className=\\"contact-form\\">
    <h2>
      Send message
    </h2>
    <Formik initialValues={{...}} onSubmit={[Function: submitForm]} validate={[Function: validate]}>
      <form onSubmit={[Function]} noValidate={true}>
        <div className=\\"contact-form_line\\">
          <FormInput label=\\"Name\\" type=\\"text\\" name=\\"name\\" onChange={[Function]} onBlur={[Function]} value=\\"\\" errors={[undefined]} touched={[undefined]} className=\\"contact-form_name\\">
            <div className=\\"form-input contact-form_name\\">
              <p className=\\"form-input_label\\">
                Name
              </p>
              <input name=\\"name\\" type=\\"text\\" onChange={[Function]} onBlur={[Function]} placeholder=\\"Name\\" value=\\"\\" test-id=\\"input-test\\" className=\\"form-input_input\\" disabled={[undefined]} />
              <div>
                <div className=\\"form-input_validation-marker undefined\\" />
                <div className=\\"form-input_error\\" />
              </div>
            </div>
          </FormInput>
        </div>
        <div className=\\"contact-form_line\\">
          <FormInput label=\\"Phone\\" type=\\"text\\" name=\\"phone\\" onChange={[Function]} onBlur={[Function]} value=\\"\\" errors={[undefined]} touched={[undefined]}>
            <div className=\\"form-input undefined\\">
              <p className=\\"form-input_label\\">
                Phone
              </p>
              <input name=\\"phone\\" type=\\"text\\" onChange={[Function]} onBlur={[Function]} placeholder=\\"Phone\\" value=\\"\\" test-id=\\"input-test\\" className=\\"form-input_input\\" disabled={[undefined]} />
              <div>
                <div className=\\"form-input_validation-marker undefined\\" />
                <div className=\\"form-input_error\\" />
              </div>
            </div>
          </FormInput>
          <FormInput label=\\"Email\\" type=\\"email\\" name=\\"email\\" onChange={[Function]} onBlur={[Function]} value=\\"\\" errors={[undefined]} touched={[undefined]}>
            <div className=\\"form-input undefined\\">
              <p className=\\"form-input_label\\">
                Email
              </p>
              <input name=\\"email\\" type=\\"email\\" onChange={[Function]} onBlur={[Function]} placeholder=\\"Email\\" value=\\"\\" test-id=\\"input-test\\" className=\\"form-input_input\\" disabled={[undefined]} />
              <div>
                <div className=\\"form-input_validation-marker undefined\\" />
                <div className=\\"form-input_error\\" />
              </div>
            </div>
          </FormInput>
        </div>
        <div className=\\"contact-form_line\\">
          <FormInput label=\\"How we can help you?\\" type=\\"text\\" name=\\"content\\" onChange={[Function]} onBlur={[Function]} value=\\"\\" errors={[undefined]} touched={[undefined]} className=\\"contact-form_message\\">
            <div className=\\"form-input contact-form_message\\">
              <p className=\\"form-input_label\\">
                How we can help you?
              </p>
              <input name=\\"content\\" type=\\"text\\" onChange={[Function]} onBlur={[Function]} placeholder=\\"How we can help you?\\" value=\\"\\" test-id=\\"input-test\\" className=\\"form-input_input\\" disabled={[undefined]} />
              <div>
                <div className=\\"form-input_validation-marker undefined\\" />
                <div className=\\"form-input_error\\" />
              </div>
            </div>
          </FormInput>
        </div>
        <FormButton label=\\"Submit\\" stylingType=\\"submit\\" type=\\"submit\\" disabled={false}>
          <button className=\\"form-button form-button_submit undefined\\" data-testid=\\"form-button\\" disabled={false} type=\\"submit\\" onClick={[undefined]}>
            Submit
          </button>
        </FormButton>
      </form>
    </Formik>
  </div>
</ContactForm>"
`;
